### PR TITLE
CIS 120 Add possibility to assert a network request was made

### PR DESCRIPTION
**Pull request by VojtaStavik**
### In this PR:
  - I made the default `URLSessionConfiguration` the client uses configurable. I had to do it to make it possible to register a custom `URLProtocol` on `Client`'s `URLSession`.

- I created `RequestRecorderURLProtocol`, a `URLProtocol` implementation which listens to all network requests and stores the last one made. You can call `getLatestRequestSynchronously()` on it to synchronously wait for a future request. This is quite cool but it makes asynchronous tests as simple to write as synchronous.

- To showcase how this can be used I added tests for `Client+Device` API calls.

Please note that this is just the first step. As we will use this more and more we can add more functionality and smooth all rough edges.

#no_changelog

---

@buh @b-onc Since this PR is big, I prepared the commit history in the way such that you can review commit by commit.

### CODE REVIEW

File: Sources/Client/Client/Client+Config.swift [hunk 1]
```diff
@@ -35,18 +37,22 @@ extension Client {
...
+        ///     - defaultURLSessionConfiguration: the URLSessionConfiguration the Client uses as
+        ///     the default configuration.
         public init(apiKey: String,
                     baseURL: BaseURL = .usEast,
                     stayConnectedInBackground: Bool = true,
                     database: Database? = nil,
                     callbackQueue: DispatchQueue? = nil,
-                    logOptions: ClientLogger.Options = []) {
+                    logOptions: ClientLogger.Options = [],
+                    defaultURLSessionConfiguration: URLSessionConfiguration = .default) {
```
Comment (b-onc) [inline Sources/Client/Client/Client+Config.swift:48]: I'm not sure if this should be configurable by the user. We're not going to be able to make some assumptions about session configuration when we make this configurable. If there's no use case for users, I'd say KISS

File: Tests/Client/RequestRecorderURLProtocol.swift [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+//
+//  Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+import XCTest
+
+/// This URLProtocol subclass allows to intercept the network communication
+/// and provides the latest network request made.
+class RequestRecorderURLProtocol: URLProtocol {
```
Comment (b-onc) [inline Tests/Client/RequestRecorderURLProtocol.swift:13]: At first I thought the `Protocol` in name refers to Swift protocols, then I started reading up and realized it refers to URL/Networking protocols 🤦 So no renaming suggestion from me (which was my initial reaction)

File: Tests/Client/RequestRecorderURLProtocol.swift [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        return false
+    }
+
+    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+        // Overriding this function is required by the superclass.
+        return request
+    }
+
+    private static var latestRequestExpectation: XCTestExpectation?
+    private static var latestRequest: URLRequest?
```
Comment (b-onc) [inline Tests/Client/RequestRecorderURLProtocol.swift:50]: I see that your style is putting variables after functions, in itself its consistent:
- static func
- overridden static func
- static variables
- instance func
- overridden instance func
- instance vars

putting vars at the end of file makes reading harder IMHO (because you could use the vars inside the functions) and it's also the opposite of what our SDK uses (vars at the top)

File: Tests/Client/RequestRecorderURLProtocol.swift [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    static func getLatestRequestSynchronously(timeout: TimeInterval) -> URLRequest? {
+        defer {
+            // Delete the used request
+            latestRequest = nil
+        }
+
+        guard latestRequest == nil else { return latestRequest }
+
+        latestRequestExpectation = .init(description: "Wait for incoming request.")
+        _ = XCTWaiter.wait(for: [latestRequestExpectation!], timeout: 1)
```
Comment (b-onc) [inline Tests/Client/RequestRecorderURLProtocol.swift:31]: `timeout` parameters is not used and hardcoded as `1`, I suggest providing default `timeout` as 1

File: Tests/Client/Unit Tests/Client+DevicesTests.swift [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    // possible at the moment.
+    lazy var client: Client = {
+        let sessionConfig = URLSessionConfiguration.default
+
+        sessionConfig.protocolClasses?.insert(RequestRecorderURLProtocol.self, at: 0)
+        Client.config = .init(apiKey: "test_api_key", defaultURLSessionConfiguration: sessionConfig)
+        return Client.shared
+    }()
+
+    func test_getDevice_createsRequest() {
```
Comment (b-onc) [inline Tests/Client/Unit Tests/Client+DevicesTests.swift:24]: This is the first time I'm seeing test naming like this and I really liked it. Definitely more readable than `testGetDeviceCreatesRequest`
Is the format:
test -> This is a test (XCTest requirement)
getDevice -> endpoint that's being tested
createsRequest -> What is the expected result
?

File: Tests/Client/Unit Tests/Client+DevicesTests.swift [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        AssertNetworkRequest(
+            method: .get,
+            path: "/devices",
+            headers: ["Content-Type": "application/json"],
+            queryParameters: ["api_key": "test_api_key"],
+            body: nil
+        )
+    }
+
+    func test_addDevice_withDeviceID_createsRequest() {
```
Comment (b-onc) [inline Tests/Client/Unit Tests/Client+DevicesTests.swift:42]: Continuing from my comment above (about test naming) I suggest naming this test as:
`test_addDeviceWithDeviceId_createsRequest` since adding a device without a device ID is not possible.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        client.devices { _ in }
+
+        // Assert
+        AssertNetworkRequest(
+            method: .get,
+            path: "/devices",
+            headers: ["Content-Type": "application/json"],
+            queryParameters: ["api_key": "test_api_key"],
+            body: nil
+        )
```
Comment (b-onc) [inline Tests/Client/Unit Tests/Client+DevicesTests.swift:39]: How about we embed the statement making the network call in this function?
```swift
func AssertNetworkRequest(
    action: @autoclosure () -> Any,
    method: Endpoint.Method,
    path: String,
    headers: [String: String]?,
    queryParameters: [String: String]?,
    body: [String: Any]?,
    timeout: TimeInterval = 0.5,
    file: StaticString = #file,
    line: UInt = #line
)
```
I think it's more inline with XCTest functions with this approach. We can use it with:
```swift
func test_getDevice_createsRequest() {
    // Setup
    let testUser = User(id: "test_user_\(UUID())")
    client.set(user: testUser, token: "test_token")

    // Assert
    AssertNetworkRequest(
        action: client.devices { _ in },
        method: .get,
        path: "/devices",
        headers: ["Content-Type": "application/json"],
        queryParameters: ["api_key": "test_api_key"],
        body: nil
    )
}
```
by adding `_ = action()` as the first line in `func AssertNetworkRequest`. The action and the expected result are coupled with this way.

**Comment by b-onc** [comment]
As a side note, this is not "meta", this is sdk:core + enchancement :) Meta would be stuff about the repo/project itself, like CI configuration, external documentation, etc. If you have seen us tagging old PRs relating to tests about meta, that's our mistake and apologizes.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    func test_addDevice_withDeviceID_createsRequest() {
+        // Setup
+        let testUser = User(id: "test_user_\(UUID())")
+        let testDeviceId = "device_id_\(UUID())"
+        client.set(user: testUser, token: "test_token")
```
Comment (b-onc) [inline Tests/Client/Unit Tests/Client+DevicesTests.swift:46]: These shared lines could be moved into `setUp`, or event in the computed property itself

File: Sources/Client/Client/Client+Config.swift (hunk 1 above)
Comment (VojtaStavik) [reply Sources/Client/Client/Client+Config.swift:48]: Good point, I didn't realize this is a user-facing config.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 5 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:24]: It's based on this tweet I saw **yesterday morning** right before starting working on this task https://twitter.com/qcoding/status/1255619642573426689?s=20 . 
First I was like ewww but then - it makes everything much more readable!

Comment (b-onc) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:24]: Same for me! "eeewwwww....wait....wow!"
Thanks for sharing the tweet!

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 7 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:39]: I understand what you're trying to achieve there. I see two reasons for not doing that:

- The request is not a result of the action but its side effect. Using it directly in the assertion implies the `action` signature is something like `() -> Request`.

- When the action is more than a couple of lines, it becomes hard to parse it visually and you'd probably refactor it into a variable anyway:
```swift
func test_getDevice_createsRequest() {
    // Setup
    let testUser = User(id: "test_user_\(UUID())")
    client.set(user: testUser, token: "test_token")

    let action = { 
       .... 
    }

    // Assert
    AssertNetworkRequest(
        action: action,
        method: .get,
        path: "/devices",
        headers: ["Content-Type": "application/json"],
        queryParameters: ["api_key": "test_api_key"],
        body: nil
    )
}
```

Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:39]: This is just the first version of this function. I'm sure we'll have a lot of improvements and ideas once we start using it extensively on day-to-day basis.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 8 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:46]: I thought about it but didn't want to overcomplicate things. I think we'll find the right pattern for this while adding more tests. But good comment 👍

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 7 above)
Comment (b-onc) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:39]: I don't agree with the points:
* Network request is a direct result of the call being made, just not the return value.
* A network request will be a result of a single call (of client, network, etc..) so if it requires more than a line, it can be boiled down a single call.

But I definitely agree that this is the first implementation and we'll improve, we can keep it like this if you think so.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 8 above)
Comment (b-onc) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:46]: Isn't keeping setup part individual for each test (although they're the same) overcomplicating?

Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:46]: I like how local everything is. You can copy-paste the function somewhere else and it will still work (mostly). I like it when tests are dumb and straightforward. When you don't need to think about the code that much and it's just there right in your face.

This is just my personal preference and I have no best-practices link I can post here. If it bothers you, feel free to add a commit and show your preferred way of doing this 👍

File: Sources/Client/Client/Client+Config.swift (hunk 1 above)
Comment (buh) [reply Sources/Client/Client/Client+Config.swift:48]: I agree with @b-onc It should be internal configuration.

File: Sources/Client/Client/Client+Request.swift [hunk 9]
```diff
@@ -16,7 +16,7 @@ extension Client {
     func setupURLSession(token: Token = "") -> URLSession {
         let headers = authHeaders(token: token)
         logger?.log(headers: headers)
-        let config = URLSessionConfiguration.default
+        let config = self.defaultURLSessionConfiguration
```
Comment (buh) [inline Sources/Client/Client/Client+Request.swift:19]: No need `self.`

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 7 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:39]: `Cancellable` is the return value of this, everything else is technically a side effect.

If you don't mind I'd keep it as is. My arguments are not strong enough to convince you, nor are your to convince me. Time will tell ... :)

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+func AssertNetworkRequest(
+    method: Endpoint.Method,
+    path: String,
+    headers: [String: String]?,
+    queryParameters: [String: String]?,
+    body: [String: Any]?,
+    timeout: TimeInterval = 0.5,
+    file: StaticString = #file,
+    line: UInt = #line
+) {
```
Comment (buh) [inline Tests/Client/Custom Assertions/AssertNetworkRequest.swift:38]: We don't start parameter from a new line.
```suggestion
func AssertNetworkRequest(method: Endpoint.Method,
                          path: String,
                          headers: [String: String]?,
                          queryParameters: [String: String]?,
                          body: [String: Any]?,
                          timeout: TimeInterval = 0.5,
                          file: StaticString = #file,
                          line: UInt = #line) {
```

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 7 above)
Comment (b-onc) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:39]: Thank you for explaining, since I don't have any more arguments let's keep it as it is 👍🏼

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    guard let requestBodyJSON = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
+        errorMessage += "\n  - Request body is not a valid JSON"
+        return
+    }
+
+    body?.forEach { (key, value) in
+        if let requestValue = requestBodyJSON[key] {
+            if String(describing: value) != String(describing: requestValue) {
```
Comment (buh) [inline Tests/Client/Custom Assertions/AssertNetworkRequest.swift:115]: I suppose the order of keys can be different. Maybe better to check this recursively.

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            errorMessage += "\n  - Missing body value for \"\(key)\""
+        }
+    }
+}
+
+private extension Array where Element == URLQueryItem {
+  /// Returns the value of the URLQueryItem with the given name. Returns `nil`
+  /// if the query item doesn't exist.
+  subscript(_ name: String) -> String? {
+    self.first(where: { $0.name == name}).flatMap({ $0.value })
```
Comment (buh) [inline Tests/Client/Custom Assertions/AssertNetworkRequest.swift:128]: No need `self.`

File: Sources/Client/Client/Client+Config.swift (hunk 1 above)
Comment (VojtaStavik) [reply Sources/Client/Client/Client+Config.swift:48]: I'll fix this once we merge https://github.com/GetStream/stream-chat-swift/pull/231 since it touches the same part of the codebase.

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        input.open()
+
+        let bufferSize = 1024
+        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
+        while input.hasBytesAvailable {
+            let read = input.read(buffer, maxLength: bufferSize)
+            if (read == 0) {
+                break  // added
+            }
+            self.append(buffer, count: read)
```
Comment (buh) [inline Tests/Client/Custom Assertions/AssertNetworkRequest.swift:145]: No need `self.`.
It's good to have this function, but currently, we don't use data from `InputStream`.

File: Tests/Client/RequestRecorderURLProtocol.swift [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+class RequestRecorderURLProtocol: URLProtocol {
+
+    /// Returns the latest network request this URLProtocol recorded.
+    ///
+    /// If no request has been made since the last time this function was invoked, it synchronously
+    /// waits for the next request to be made.
+    ///
+    /// - Parameter timeout: Specifies the time the function waits for a new request
+    /// to be made.
+    static func getLatestRequestSynchronously(timeout: TimeInterval) -> URLRequest? {
```
Comment (buh) [inline Tests/Client/RequestRecorderURLProtocol.swift:22]: We usually don't use `get` and `set` words. It's not swifty :)

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 8 above)
Comment (b-onc) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:46]: I see, it's a tradeoff between locality/ease of understanding vs less repetition. Let's keep it as it is for this PR 👍🏼
Disclosure: when I write tests, I'll use `setUp` function 😂

File: Tests/Client/RequestRecorderURLProtocol.swift (hunk 4 above)
Comment (buh) [reply Tests/Client/RequestRecorderURLProtocol.swift:31]: Still doesn't use `timeout`.

File: Tests/Client/RequestRecorderURLProtocol.swift [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        return latestRequest
+    }
+
+    override class func canInit(with request: URLRequest) -> Bool {
+        DispatchQueue.main.async {
+            latestRequest = request
+            latestRequestExpectation?.fulfill()
+
+        }
+        return false
```
Comment (buh) [inline Tests/Client/RequestRecorderURLProtocol.swift:41]:
```suggestion
        }
        
        return false
```

File: Tests/Client/RequestRecorderURLProtocol.swift (hunk 3 above)
Comment (buh) [reply Tests/Client/RequestRecorderURLProtocol.swift:50]: I agree with @b-onc 
Please put vars on top and functions after. Static stuff first, instance after.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 5 above)
Comment (buh) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:24]: I can't say I like it, but I'll agree with you if it useful for tests.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 7 above)
Comment (buh) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:39]: Both suggestions look fine for me.

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 6 above)
Comment (buh) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift:42]: I agree with @b-onc

File: Tests/Client/RequestRecorderURLProtocol.swift (hunk 4 above)
Comment (VojtaStavik) [reply Tests/Client/RequestRecorderURLProtocol.swift:31]: @buh That isn't that surprising since there wasn't any new commit in the branch in the last 23 hours, hm? :trollface:

Comment (buh) [reply Tests/Client/RequestRecorderURLProtocol.swift:31]: I saw the `timeout` parameter and thought you added, but forgot to use it.

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift (hunk 11 above)
Comment (VojtaStavik) [reply Tests/Client/Custom Assertions/AssertNetworkRequest.swift:115]: The order of the keys doesn't matter in this implementation.

File: Tests/Client/RequestRecorderURLProtocol.swift (hunk 15 above)
Comment (VojtaStavik) [reply Tests/Client/RequestRecorderURLProtocol.swift]: fixed

File: Tests/Client/RequestRecorderURLProtocol.swift (hunk 14 above)
Comment (VojtaStavik) [reply Tests/Client/RequestRecorderURLProtocol.swift]: fixed

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift (hunk 13 above)
Comment (VojtaStavik) [reply Tests/Client/Custom Assertions/AssertNetworkRequest.swift]: - fixed
- it's used actually. the body of the tests in the PR comes from `request.httpBodyStream`. Is it unexpected @buh ?

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift (hunk 12 above)
Comment (VojtaStavik) [reply Tests/Client/Custom Assertions/AssertNetworkRequest.swift]: fixed

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift (hunk 10 above)
Comment (VojtaStavik) [reply Tests/Client/Custom Assertions/AssertNetworkRequest.swift]: fixed

File: Sources/Client/Client/Client+Request.swift (hunk 9 above)
Comment (VojtaStavik) [reply Sources/Client/Client/Client+Request.swift]: fixed

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 8 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift]: I refactored the lines shared by all tests cases to the `setUp` function

File: Tests/Client/Unit Tests/Client+DevicesTests.swift (hunk 6 above)
Comment (VojtaStavik) [reply Tests/Client/Unit Tests/Client+DevicesTests.swift]: fixed

File: Tests/Client/RequestRecorderURLProtocol.swift (hunk 4 above)
Comment (VojtaStavik) [reply Tests/Client/RequestRecorderURLProtocol.swift]: fixed

File: Tests/Client/RequestRecorderURLProtocol.swift (hunk 3 above)
Comment (VojtaStavik) [reply Tests/Client/RequestRecorderURLProtocol.swift]: fixed

File: Sources/Client/Client/Client+Config.swift (hunk 1 above)
Comment (VojtaStavik) [reply Sources/Client/Client/Client+Config.swift]: fixed

File: Tests/Client/Custom Assertions/AssertNetworkRequest.swift (hunk 13 above)
Comment (buh) [reply Tests/Client/Custom Assertions/AssertNetworkRequest.swift]: Yes, it's weird.
